pkg/util/push/request.go

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type Request struct {
+	cleanups []func()
```

Comment:
To avoid resizing this slice when adding cleanups, we could either preallocate it, or perhaps even use a big enough array (then we need extra field to know how many cleanup functions we have).

Reply:
we currently call add cleanup up to 4 times in the distributor. Preallocated it with size 10